### PR TITLE
Fix flakiness by synchronising the State

### PR DESCRIPTION
**Pull request by Pldi23**
Description:

Cloudbees CI reported a flake in 
[org.jenkinsci.plugins.workflow.support.steps.StageStepTest.serializability](https://github.com/jenkinsci/pipeline-stage-step-plugin/blob/master/src/test/java/org/jenkinsci/plugins/workflow/support/steps/StageStepTest.java#L173-L200)

Error Log:
[stderr.txt](https://github.com/jenkinsci/workflow-support-plugin/files/9150162/stderr.txt)

How to reproduce:
I wasn't managed to reproduce it locally without code changes, but if we add `TimeUnit.MILLISECONDS.sleep(200);` before this [line](https://github.com/jenkinsci/workflow-support-plugin/blob/master/src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java#L110) the StageStepTest.serializability started to flake regularly.

Assumption:
My assumption that there is a deadlock possibility.

Solution:
As a solution I want to suggest to synchronize the State local variable (the same synchronisation mechanism is used [here](https://github.com/jenkinsci/workflow-support-plugin/blob/master/src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java#L141)) 

- [ ] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [ ] Ensure that the pull request title represents the desired changelog entry
- [ ] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java
Comment (Pldi23) [inline src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java:124]: I did not notice some tests that failed because of the race condition here, but since we are synchronizing the success method, then we need to synchronize here too.

**Review by jmdesprez — APPROVED** [review]

**Comment by Pldi23** [comment]
Waiting for a review from maintainers

**Comment by Pldi23** [comment]
@jglick could you please review

File: src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java:109]: I am somewhat concerned about holding a lock while calling `onSuccess` here. Is it effective to hold the lock on `State` only while accessing its fields? For example

```java
public static void success(String k, Object returnValue) {
    State s = State.get();
    synchronized (s) {
        if (!s.contexts.containsKey(k)) {
            System.err.println("Planning to unblock " + k + " as success");
            s.returnValues.put(k, returnValue);
            return;
        }
    }
    System.err.println("Unblocking " + k + " as success");
    getContext(s, k).onSuccess(returnValue);
}
```

**Review by jglick — COMMENTED** [review]
IIUC to match https://github.com/jenkinsci/workflow-support-plugin/blob/a1c71061486b4ce3db98576f9435f7ca054b6aec/src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java#L141-L148

File: src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java
Comment (Pldi23) [reply src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java]: Yeah that's even better.

**Review by jglick — APPROVED** [review]

**Pull request merged**
